### PR TITLE
Docs: Explain how `touch` lacks dirty-tracking

### PR DESCRIPTION
**Pull request by jaredbeck**
:+1: by @y-yagi in https://github.com/rails/rails/issues/33429

[ci skip]

### CODE REVIEW

File: activerecord/lib/active_record/persistence.rb [hunk 1]
```diff
@@ -621,8 +621,12 @@ def reload(options = nil)
 
     # Saves the record with the updated_at/on attributes set to the current time
     # or the time specified.
-    # Please note that no validation is performed and only the +after_touch+,
-    # +after_commit+ and +after_rollback+ callbacks are executed.
+    #
+    # - No validation is performed.
```
Comment (gmcgibbon) [inline activerecord/lib/active_record/persistence.rb]: Looking at other lists in the docs, it appears `*` should be used for delimiting lists. Can you change your `-` to `*`s? 😄

File: activerecord/lib/active_record/persistence.rb [hunk 2]
```diff
@@ -621,8 +621,12 @@ def reload(options = nil)
...
+    # - Only the +after_touch+, +after_commit+ and +after_rollback+ callbacks
+    #   are executed.
+    # - No dirty-tracking occurs. Methods like `changed?`, `previous_changes`,
+    #   and `changes` return the same values before and after `touch`.
```
[the remark this answers is not in the document]
Comment (jaredbeck) [reply activerecord/lib/active_record/persistence.rb]: > No dirty-tracking occurs. Methods like `changed?`, `previous_changes`, and `changes` return the same values before and after `touch`.

Please see the following, which I think demonstrates this:

```ruby
# user.rb
  after_touch(
    lambda do
      puts format('after_touch: changed?: %s', changed?)
      puts format('after_touch: previous_changes: %s', previous_changes)
      puts format('after_touch: changes: %s', changes)
    end
  )

# console
u = User.find 1
u.changes #=> {}
u.previous_changes #=> {}
u.changed? #=> false
u.touch
# UPDATE `users` SET `updated_at` ...
# after_touch: changed?: false
# after_touch: previous_changes: {}
# after_touch: changes: {}
u.changes #=> {}
u.previous_changes #=> {}
u.changed? #=> false
```

[the remark this answers is not in the document]
Comment (gmcgibbon) [reply activerecord/lib/active_record/persistence.rb]: I see empty `changes` as well with this script: https://gist.github.com/gmcgibbon/35353ef08bdc9ab5c0c4727dddc0d754

**Review by gmcgibbon — APPROVED** [review]

File: activerecord/lib/active_record/persistence.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (kamipo) [reply activerecord/lib/active_record/persistence.rb]: I think that these scripts doesn't check the effect of `clear_attribute_change`.

[the remark this answers is not in the document]
Comment (gmcgibbon) [reply activerecord/lib/active_record/persistence.rb]: @kamipo If I understand correctly, you mean when the record is already dirty and then touched, right?

I see now that it clears previous dirty tracking of the touched column. We should change the wording of this point to say that previous changes are forgotten.

**Comment by jaredbeck** [comment]
> @kamipo If I understand correctly, you mean when the record is already dirty and then touched, right? I see now that it clears previous dirty tracking of the touched column. We should change the wording of this point to say that previous changes are forgotten.

I've added a sentence to that effect.

File: activerecord/lib/active_record/persistence.rb [hunk 3]
```diff
@@ -621,8 +621,14 @@ def reload(options = nil)
...
     # Saves the record with the updated_at/on attributes set to the current time
     # or the time specified.
-    # Please note that no validation is performed and only the +after_touch+,
-    # +after_commit+ and +after_rollback+ callbacks are executed.
+    #
+    # * No validation is performed.
+    # * Only the +after_touch+, +after_commit+ and +after_rollback+ callbacks
+    #   are executed.
+    # * The dirty-tracking state is cleared, so unpersisted changes are no
+    #   longer tracked by methods like `changed?`, `previous_changes`,
```
Comment (y-yagi) [inline activerecord/lib/active_record/persistence.rb]: The backticks can't use for code font in RDoc.  
Please change to use `+` just like the previous line.

**Comment by jaredbeck** [comment]
> The backticks can't use for code font in RDoc. Please change to use + just like the previous line.

Changed.

File: activerecord/lib/active_record/persistence.rb [hunk 4]
```diff
@@ -621,8 +621,14 @@ def reload(options = nil)
...
     # or the time specified.
-    # Please note that no validation is performed and only the +after_touch+,
-    # +after_commit+ and +after_rollback+ callbacks are executed.
+    #
+    # * No validation is performed.
+    # * Only the +after_touch+, +after_commit+ and +after_rollback+ callbacks
+    #   are executed.
+    # * The dirty-tracking state is cleared, so unpersisted changes are no
+    #   longer tracked by methods like +changed?+, +previous_changes+,
+    #   and +changes+. Then, there is no dirty-tracking of the +touch+ itself,
```
Comment (kamipo) [inline activerecord/lib/active_record/persistence.rb]: This sentence "The dirty-tracking state is cleared, so unpersisted changes are no longer tracked" still makes misreading, the `touch` doesn't clear *all* unpersisted changes.

Comment (jaredbeck) [reply activerecord/lib/active_record/persistence.rb]: > the touch doesn't clear all unpersisted changes.

Touch only clears unpersisted changes to the affected timestamps?

**Comment by jaredbeck** [comment]
Need feedback please, not sure what changes to make. Commits welcome, thanks!

File: activerecord/lib/active_record/persistence.rb (hunk 4 above)
Comment (gmcgibbon) [reply activerecord/lib/active_record/persistence.rb]: @jaredbeck Essentially, yes. It clears changes for any column you are touching. Can you reword it to "Dirty-tracking state is cleared *for touched attributes*, so ..."? See [my updated test script](https://gist.github.com/gmcgibbon/35353ef08bdc9ab5c0c4727dddc0d754).

**Comment by jaredbeck** [comment]
> @jaredbeck Essentially, yes. It clears changes for any column you are touching. Can you reword it to "Dirty-tracking state is cleared _for touched attributes_, so ..."?

I have made this change, thanks.

File: activerecord/lib/active_record/persistence.rb [hunk 5]
```diff
@@ -622,8 +622,15 @@ def reload(options = nil)
...
     # Saves the record with the updated_at/on attributes set to the current time
     # or the time specified.
-    # Please note that no validation is performed and only the +after_touch+,
-    # +after_commit+ and +after_rollback+ callbacks are executed.
+    #
+    # * No validation is performed.
+    # * Only the +after_touch+, +after_commit+ and +after_rollback+ callbacks
+    #   are executed.
+    # * The dirty-tracking state is cleared for touched attributes, so
+    #   unpersisted changes are no longer tracked by methods like +changed?+,
```
Comment (kamipo) [inline activerecord/lib/active_record/persistence.rb]: The dirty-tracking state is cleared only for touched (= persisted) attributes, so "so unpersisted changes are no longer tracked by methods like ..." part is not true or makes misreading.

Comment (jaredbeck) [reply activerecord/lib/active_record/persistence.rb]: Can you suggest an alternative wording please? I'm on at least my fourth revision here (all good stuff) so, let's just get it done. Thanks!

**Comment by jaredbeck** [comment]
>  kamipo closed this in d1107f4 13 hours ago 

Thanks! Fixing https://github.com/rails/rails/issues/33429 is even better than improving the docs. :heart:
